okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -106,10 +106,16 @@ public HttpCodec newStream(
     boolean connectionRetryEnabled = client.retryOnConnectionFailure();
 
     try {
+      RealConnection existingConnection = connection;
+
       RealConnection resultConnection = findHealthyConnection(connectTimeout, readTimeout,
           writeTimeout, connectionRetryEnabled, doExtensiveHealthChecks);
       HttpCodec resultCodec = resultConnection.newCodec(client, chain, this);
 
+      if (existingConnection != connection) {
```

Comment:
Not a perfect example, but shows that you need to look at the inet address in order to cleanly ignore a connection found that you already traced in start/end.

```
public class CEL {
  public static void main(String[] args) throws IOException {
    OkHttpClient client = new OkHttpClient.Builder().eventListener(new EventListener() {
      @Override
      public void connectStart(Call call, InetSocketAddress inetSocketAddress, Proxy proxy) {
        System.out.println("connectStart " + inetSocketAddress);
      }

      @Override public void connectEnd(Call call, InetSocketAddress inetSocketAddress,
          @Nullable Protocol protocol, @Nullable Throwable throwable) {
        System.out.println("connectEnd " + inetSocketAddress + " " + throwable);
      }

      @Override public void connectionFound(Call call, Connection connection) {
        System.out.println("connectionFound " + connection.route().inetSocketAddress);
      }
    }).build();

    client.newCall(new Request.Builder().url("https://google.com/robots.txt").build())
        .execute().body().string();

    System.out.println("---");

    client.newCall(new Request.Builder().url("http://google.com/robots.txt").build())
        .execute().body().string();

    System.out.println("---");

    client.newCall(new Request.Builder().url("https://google.com/robots.txt").build())
        .execute().body().string();
  }
}
```

```
connectStart google.com/216.58.195.78:443
connectEnd google.com/216.58.195.78:443 null
connectionFound google.com/216.58.195.78:443
connectStart www.google.com/172.217.5.100:443
connectEnd www.google.com/172.217.5.100:443 null
connectionFound www.google.com/172.217.5.100:443
---
connectStart google.com/216.58.195.78:80
connectEnd google.com/216.58.195.78:80 null
connectionFound google.com/216.58.195.78:80
connectionFound www.google.com/172.217.5.100:443
---
connectionFound google.com/216.58.195.78:443
connectionFound www.google.com/172.217.5.100:443
```